pkg/client/event_tracking.go

```diff
@@ -0,0 +1,138 @@
+package client
+
+import (
+	"encoding/hex"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// keyGenKeepTrack is used to track key generation after keep opened event is
+// received. It is used to ensure that the process execution is not duplicated,
+// e.g. when the client receives the same event multiple times. When event is
+// received, it should be noted in this struct. When the signer generation
+// process completes (no matter if it succeeded or failed), it should be removed
+// from this struct.
+type keyGenKeepTrack struct {
+	*keepEventTrack
+}
+
+// closeKeepTrack is used to track keep close events. It is used to ensure
+// that the process execution is not duplicated, e.g. when the client receives
+// the same event multiple times. When event is received, it should be noted
+// in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type closeKeepTrack struct {
+	*keepEventTrack
+}
+
+// terminateKeepTrack is used to track keep terminate events. It is used to
+// ensure that the process execution is not duplicated, e.g. when the client
+// receives the same event multiple times. When event is received, it should be
+// noted in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type terminateKeepTrack struct {
+	*keepEventTrack
+}
+
+// keepEventTrack is a simple event track implementation allowing to track
+// events that are happening once in the entire keep history, for example:
+// - keep opened, generating signer
+// - keep getting closed
+// - keep getting archived
+//
+// Those events do not require any special tracking or additional fields to
+// distinguish them. It is used to ensure that the process execution is not
+// duplicated, e.g. when the client receives the same event multiple times. When
+// event is received, it should be noted in this struct. When the signer
+// generation process completes (no matter if it succeeded or failed), it should
+// be removed from this struct.
+type keepEventTrack struct {
+	data  map[string]bool // <keep, bool>
+	mutex *sync.Mutex
```

Comment:
Why do we use a mutex pointer here?

Reply:
I just copied this code without changing this piece but... let's take this as an opportunity and fix it.

Reply:
Done in 7588bc22432c9a969c5a37a81a51a305dbf062eb.